File: api/tools.go
package api

import (
	"context"
	"net/http"
	"net/url"
	"time"
)

// Tool-level governance and the integrity quarantine
// (docs/modules/controlplane.md: GET /v1/tools, PUT
// /v1/tools/{server}/{tool}, GET /v1/quarantine, DELETE
// /v1/quarantine/{exposed}).
//
// Both stores live in <data>/state, NOT in the registry — the same files the
// gateway and the daemon consult, so a decision taken here is the decision
// the call plane enforces. A daemon assembled without a state directory
// answers the uniform 404, which a frontend renders as "unavailable on this
// daemon" (IsCode + ErrCodeNotFound), never as an empty list.

// Tool is one tool's governance state, keyed by (server, RAW tool name).
//
// Raw names are the key on purpose: a state keyed on the EXPOSED name would
// move out from under itself the moment an override renamed the tool.
type Tool struct {
	Server string `json:"server"`
	Tool   string `json:"tool"`
	// Status is the integrity approval state ("" when no record exists
	// yet). It is ORTHOGONAL to Disabled: switching a tool off and back on
	// never discards an approval, and never grants one.
	Status   string `json:"status,omitempty"`
	Disabled bool   `json:"disabled"`
	// ApprovedHash and CurrentHash are the fingerprint pin and what the
	// live definition hashes to now; a difference is drift.
	ApprovedHash string `json:"approved_hash,omitempty"`
	CurrentHash  string `json:"current_hash,omitempty"`
	// OverrideName and OverrideDescription are the stored presentation
	// overrides, empty when none is set.
	OverrideName        string `json:"override_name,omitempty"`
	OverrideDescription string `json:"override_description,omitempty"`
}

// Drifted reports that the live definition no longer matches the approved
// fingerprint. Both hashes must be known: an unknown one is "we cannot
// tell", which must not read as "unchanged".
func (t Tool) Drifted() bool {
	return t.ApprovedHash != "" && t.CurrentHash != "" && t.ApprovedHash != t.CurrentHash
}

// ToolOverride is an edit of one tool's local presentation. A nil field is
// left untouched, so blanking a description cannot silently drop a rename.
//
// Description is the NEUTRALIZATION path for a prompt-injection carrier: the
// downstream keeps its poisoned description, agenthub simply stops
// forwarding it. Name replaces the RAW name before namespacing, i.e. it
// changes what a client sees.
type ToolOverride struct {
	Name        *string `json:"name,omitempty"`
	Description *string `json:"description,omitempty"`
	// Clear removes the override entirely. It is exclusive with the two
	// field edits (the daemon refuses the combination rather than guessing
	// an order).
	Clear bool `json:"clear,omitempty"`
}

// ToolPatch is the body of a tool-governance write. A nil field is left
// untouched, so a kill-switch flip never disturbs an override and vice
// versa. The override fields are flat rather than nested because the daemon
// applies them to a different store than Enabled, in a fixed order: a
// disable first, an enable last, so a failure between the two stores cannot
// land in the looser half-state.
type ToolPatch struct {
	// Enabled false is the kill switch.
	Enabled *bool `json:"enabled,omitempty"`
	// OverrideName replaces the raw name before namespacing.
	OverrideName *string `json:"override_name,omitempty"`
	// OverrideDescription replaces the downstream description verbatim.
	OverrideDescription *string `json:"override_description,omitempty"`
	// ClearOverride drops the override entirely; exclusive with the two
	// field edits above (the daemon refuses the combination rather than
	// guessing an order).
	ClearOverride bool `json:"clear_override,omitempty"`
}

// ToolList is the answer to Tools.List.
type ToolList struct {
	Generation uint64 `json:"generation"`
	Tools      []Tool `json:"tools"`
}

// ToolWrite is what a tool-governance write returns.
type ToolWrite struct {
	WriteResult
	Server string `json:"server"`
	Tool   string `json:"tool"`
	// Enabled is present when the write touched the kill switch.
	Enabled *bool  `json:"enabled,omitempty"`
	Status  string `json:"status,omitempty"`
	// OverrideCleared and Override report the override half of the write.
	OverrideCleared bool               `json:"override_cleared,omitempty"`
	Override        *ToolOverrideValue `json:"override,omitempty"`
}

// ToolOverrideValue is a STORED override as read back: plain strings, where
// the edit form (ToolOverride) uses pointers to tell "leave alone" from
// "set to empty".
type ToolOverrideValue struct {
	Name        string `json:"name,omitempty"`
	Description string `json:"description,omitempty"`
}

// ToolsService is the tool-level governance surface: the kill switch and the
// presentation overrides.
//
// Everything here works OFFLINE from a live connection: an operator must be
// able to disable a suspicious tool without first starting the server that
// serves it. That is the whole point of a kill switch.
//
// Precondition caveat: these stores are NOT the registry, so their writes do
// not move the registry generation. expectedGeneration is still honored, but
// as an advisory check ("your view is stale"), not as a compare-and-swap on
// these files — their own cross-process locks serialize those.
type ToolsService struct{ c *Client }

// List returns the governance state of every known tool. A non-empty server
// narrows the listing to that server.
func (s *ToolsService) List(ctx context.Context, server string) (ToolList, error) {
	var q url.Values
	if server != "" {
		q = url.Values{"server": {server}}
	}
	var out ToolList
	err := s.c.do(ctx, http.MethodGet, "/tools", q, nil, &out)
	return out, err
}

// SetEnabled flips the operator kill switch for one tool.
func (s *ToolsService) SetEnabled(
	ctx context.Context, server, tool string, enabled bool, expectedGeneration uint64,
) (ToolWrite, error) {
	return s.patch(ctx, server, tool, ToolPatch{Enabled: &enabled}, expectedGeneration)
}

// SetOverride sets or clears one tool's presentation override.
func (s *ToolsService) SetOverride(
	ctx context.Context, server, tool string, ov ToolOverride, expectedGeneration uint64,
) (ToolWrite, error) {
	return s.patch(ctx, server, tool, ToolPatch{
		OverrideName:        ov.Name,
		OverrideDescription: ov.Description,
		ClearOverride:       ov.Clear,
	}, expectedGeneration)
}

func (s *ToolsService) patch(
	ctx context.Context, server, tool string, patch ToolPatch, expectedGeneration uint64,
) (ToolWrite, error) {
	var out ToolWrite
	err := s.c.doWrite(ctx, http.MethodPut,
		"/tools/"+url.PathEscape(server)+"/"+url.PathEscape(tool), nil, expectedGeneration, patch, &out)
	return out, err
}

// QuarantineEntry is one quarantined tool, keyed by the CLIENT-VISIBLE
// exposed name: that is what an agent could have called, so that is what the
// quarantine tracks.
type QuarantineEntry struct {
	Exposed     string `json:"exposed"`
	Server      string `json:"server"`
	Tool        string `json:"tool"`
	Reason      string `json:"reason,omitempty"`
	PinnedHash  string `json:"pinned_hash,omitempty"`
	CurrentHash string `json:"current_hash,omitempty"`
	// At is when the tool was isolated (UTC, RFC 3339 on the wire).
	At time.Time `json:"at"`
}

// QuarantineRelease is the outcome of releasing one entry: the write tail
// plus the entry that was released, so a frontend can report exactly what it
// just un-isolated instead of re-reading to find out.
type QuarantineRelease struct {
	WriteResult
	Exposed string          `json:"exposed"`
	Entry   QuarantineEntry `json:"entry"`
	// Released is true when an entry was actually removed. An exposed name
	// that is not in the set answers 404 rather than a cheerful success: a
	// typo must not report "released" while the real quarantine stands.
	Released bool `json:"released"`
}

// QuarantineList is the answer to Quarantine.List.
type QuarantineList struct {
	Generation uint64            `json:"generation"`
	Entries    []QuarantineEntry `json:"entries"`
}

// QuarantineService reads and releases the integrity quarantine.
type QuarantineService struct{ c *Client }

// List returns every quarantined tool.
func (s *QuarantineService) List(ctx context.Context) (QuarantineList, error) {
	var out QuarantineList
	err := s.c.do(ctx, http.MethodGet, "/quarantine", nil, nil, &out)
	return out, err
}

// Release takes one tool out of quarantine, addressed by its EXPOSED name —
// the human re-approve step.
//
// A name that is not in the set answers ErrCodeNotFound instead of
// succeeding, so a typo cannot look like a release that happened.
func (s *QuarantineService) Release(
	ctx context.Context, exposed string, expectedGeneration uint64,
) (QuarantineRelease, error) {
	var out QuarantineRelease
	err := s.c.doWrite(ctx, http.MethodDelete, "/quarantine/"+url.PathEscape(exposed), nil,
		expectedGeneration, nil, &out)
	return out, err
}

File: cmd/agenthub-gui/frontend/src/pages/tools.ts
// Tools page: the per-tool kill switch, the presentation overrides and the
// integrity quarantine.
//
// Everything here works OFFLINE from a live connection: an operator must be
// able to disable a suspicious tool without first starting the server that
// serves it. That is the whole point of a kill switch, and it is why the
// listing comes from the governance store rather than from a tool discovery.
//
// Keys are RAW downstream tool names, never the exposed ones — a state keyed
// on the exposed name would move out from under itself the moment an override
// renamed the tool. The quarantine is the exception and is keyed by the
// EXPOSED name on purpose: that is the name an agent could have called.
//
// The description override is the NEUTRALIZATION path for a prompt-injection
// carrier: the downstream keeps its poisoned description, agenthub simply
// stops forwarding it.

import { hub } from "../bridge";
import { clear, el, empty, relTime, section, table } from "../dom";
import type { Page } from "../page";
import { failureState, noticeSlot, runWrite } from "../page";
import { FILTER_BLOCKS_GLOBAL, button, confirmAction, controls, field, formHost, textInput } from "../ui";
import type { QuarantineEntry, QuarantineList, Tool, ToolList } from "../types";
import { toolDrifted } from "../types";

// ---------------------------------------------------------------------------
// The quarantine alert and its dismissal scope (docs/modules/gui.md)
// ---------------------------------------------------------------------------

/**
 * Dismissing a quarantine alert stores a CONTENT SIGNATURE, never a boolean.
 *
 * A boolean "the user has seen the quarantine alert" is wrong in the one case
 * that matters. The operator looks at a drifted tool, decides it can wait,
 * and hides the notice. Two days later the same tool drifts again — new
 * definition, new hash, new timestamp, genuinely new information — and a
 * boolean keeps it hidden, because as far as it knows this notice was already
 * dealt with. The signature makes "dealt with" mean what the user actually
 * meant: dealt with THIS, and anything new comes back.
 *
 * The signature covers the exposed name, the origin, the current hash and the
 * timestamp, i.e. everything a re-quarantine would change. It deliberately
 * does not cover the reason text, so a wording change downstream cannot
 * resurrect an alert on its own.
 */
const DISMISS_KEY = "agenthub.quarantine.dismissed";

/** FNV-1a, 32 bit. Not a security primitive and not used as one: this is a
 *  cache key for "is this the same notice", and collisions cost at worst one
 *  alert that stays hidden until the next change. */
function signatureOf(entries: QuarantineEntry[]): string {
  const canon = entries
    // \u0000 as the field separator, written escaped. The same three
    // bytes used to sit in this file literally, which made git classify
    // the whole page as binary: no diff, no blame, and grep skipping it
    // entirely. A separator that cannot occur in the data is the right
    // idea; spelling it in raw bytes is not.
    .map((q) => `${q.exposed}\u0000${q.server}/${q.tool}\u0000${q.current_hash ?? ""}\u0000${q.at}`)
    .sort()
    .join("\n");
  let h = 0x811c9dc5;
  for (let i = 0; i < canon.length; i++) {
    h ^= canon.charCodeAt(i);
    h = Math.imul(h, 0x01000193) >>> 0;
  }
  return h.toString(16).padStart(8, "0");
}

function dismissedSignature(): string {
  try {
    return localStorage.getItem(DISMISS_KEY) ?? "";
  } catch {
    return "";
  }
}

function dismissSignature(sig: string): void {
  try {
    localStorage.setItem(DISMISS_KEY, sig);
  } catch {
    // Storage unavailable: the alert simply reappears on the next visit,
    // which is the fail-toward-visibility direction.
  }
}

export function toolsPage(): Page {
  let root: HTMLElement | null = null;
  const slot = noticeSlot();
  const form = formHost();
  let tools: ToolList | null = null;
  let quarantine: QuarantineList | null = null;
  let filter = "";

  const generation = (): number => tools?.generation ?? 0;

  function overrideForm(t: Tool): Node {
    const name = textInput(t.override_name ?? "", t.tool);
    const description = el("textarea", { class: "input textarea" }) as HTMLTextAreaElement;
    description.value = t.override_description ?? "";
    description.rows = 4;
    description.placeholder = "replaces the downstream description verbatim";

    const apply = button("Apply override", "btn", () => {
      void runWrite(
        slot,
        () => draw(),
        (r) => `Override applied to ${r.server}/${r.tool}.`,
        () =>
          hub.setToolOverride(
            t.server,
            t.tool,
            { name: name.value.trim(), description: description.value },
            generation(),
          ),
      ).then((ok) => {
        if (ok) form.hide();
      });
    });
    const drop = button("Clear override", "btn btn-secondary", () => {
      void runWrite(
        slot,
        () => draw(),
        (r) => `Override cleared on ${r.server}/${r.tool}.`,
        () => hub.setToolOverride(t.server, t.tool, { clear: true }, generation()),
      ).then((ok) => {
        if (ok) form.hide();
      });
    });

    return el("div", { class: "panel panel-inset" }, [
      el("h3", { text: `Override ${t.server}/${t.tool}` }),
      field("Exposed name", name, "replaces the raw name before namespacing — this is what a client sees"),
      field("Description", description, "the neutralization path: the downstream keeps its text, agenthub stops forwarding it"),
      controls(apply, drop, button("Cancel", "btn btn-secondary", () => form.hide())),
    ]);
  }

  async function toggle(t: Tool): Promise<void> {
    const next = t.disabled;
    if (!next) {
      const ok = await confirmAction({
        title: `Disable ${t.server}/${t.tool}?`,
        body: "The tool stops being callable by every client immediately.",
        consequences: [
          "The kill switch is independent of the integrity approval: switching a tool off and back on neither discards an approval nor grants one.",
        ],
        confirmLabel: "Disable",
      });
      if (!ok) return;
    }
    await runWrite(
      slot,
      () => draw(),
      (r) => `${r.server}/${r.tool} ${next ? "enabled" : "disabled"}.`,
      () => hub.setToolEnabled(t.server, t.tool, next, generation()),
    );
  }

  async function release(exposed: string): Promise<void> {
    const ok = await confirmAction({
      title: `Re-approve ${exposed}?`,
      body: "The tool becomes callable again under the definition it has right now.",
      consequences: [
        "This is the human re-approve step: the definition that tripped the integrity check is the one you are accepting, and it becomes the new baseline.",
        "Nothing else is re-approved: other quarantined tools, and this tool's next change, still stop here.",
      ],
      confirmLabel: "Re-approve",
      danger: true,
      cli: `agenthub tool quarantine release ${exposed}`,
    });
    if (!ok) return;
    await runWrite(
      slot,
      () => draw(),
      (r) => `${r.exposed} released from quarantine.`,
      () => hub.releaseQuarantine(exposed, quarantine?.generation ?? 0),
    );
  }

  /**
   * The alert card above the quarantine listing, or the resting-state card
   * when there is nothing to report.
   *
   * The resting card (docs/modules/gui.md) costs one line and buys the
   * thing this product is otherwise structurally bad at: protection nobody
   * ever sees produces no trust. A user who has never had a tool tampered
   * with has, from the interface alone, no evidence that anything is
   * watching — and will conclude, correctly for all they can tell, that
   * nothing is.
   */
  function quarantineAlert(entries: QuarantineEntry[]): Node {
    if (entries.length === 0) {
      return el("div", { class: "resting" }, [
        el("span", { class: "dot neutral" }),
        el("span", {
          text: "Watching for tool tampering, description poisoning and injected output. Nothing is wrong right now.",
        }),
      ]);
    }

    const signature = signatureOf(entries);
    if (dismissedSignature() === signature) {
      // Dismissed for THIS content. The listing below and the sidebar
      // counter both still show it: dismissing is "stop shouting", never
      // "stop reporting".
      return el("p", {
        class: "hint",
        text: `${entries.length} quarantined ${entries.length === 1 ? "tool is" : "tools are"} listed below. The alert is hidden until something changes.`,
      });
    }

    const filtered = filter ? entries.filter((q) => q.server === filter) : entries;
    const keepBlocked = button("Keep blocked", "btn btn-secondary", () => {
      dismissSignature(signature);
      void draw();
    });
    if (filter) {
      // docs/modules/gui.md: a global action under a filter would dismiss
      // alerts for servers this view is not showing.
      keepBlocked.disabled = true;
      keepBlocked.title = FILTER_BLOCKS_GLOBAL;
    }

    return el("div", { class: "alert", role: "alert" }, [
      el("header", {}, [
        el("span", {
          class: "alert-title",
          text: `${entries.length} ${entries.length === 1 ? "tool was" : "tools were"} isolated by the integrity check`,
        }),
        keepBlocked,
      ]),
      el("span", {
        text:
          "These tools are blocked right now — no client can call them. Re-approve one only if you " +
          "know why its definition changed.",
      }),
      ...filtered.map((q) =>
        el("div", { class: "alert-item" }, [
          el("strong", { text: q.exposed }),
          el("span", { class: "meta", text: `${q.server}/${q.tool} · ${q.reason || "changed"} · ${relTime(q.at)}` }),
          q.pinned_hash ? el("div", { class: "meta mono", text: `approved ${q.pinned_hash}` }) : null,
          q.current_hash ? el("div", { class: "meta mono", text: `current  ${q.current_hash}` }) : null,
          controls(button("Re-approve", "btn btn-deny", () => void release(q.exposed))),
        ]),
      ),
      filter && filtered.length !== entries.length
        ? el("p", {
            class: "hint",
            text: `${entries.length - filtered.length} more quarantined ${entries.length - filtered.length === 1 ? "tool is" : "tools are"} hidden by the “${filter}” filter.`,
          })
        : null,
      el("p", {
        class: "hint",
        text: "“Keep blocked” hides this alert for exactly these tools and this definition. If any of them changes again, the alert comes back — and the counter in the sidebar never goes away while anything is quarantined.",
      }),
    ]);
  }

  function toolRow(t: Tool): (Node | string)[] {
    const drifted = toolDrifted(t);
    return [
      el("div", {}, [
        el("strong", { text: `${t.server}/${t.tool}` }),
        t.override_name ? el("div", { class: "muted", text: `exposed as ${t.override_name}` }) : null,
        t.override_description
          ? el("div", { class: "muted", text: "description overridden" })
          : null,
      ]),
      el("div", {}, [
        t.disabled
          ? el("span", { class: "badge badge-disabled", text: "disabled" })
          : el("span", { class: "badge badge-healthy", text: "enabled" }),
        t.status ? el("div", { class: "muted", text: t.status }) : null,
      ]),
      drifted
        ? el("div", {}, [
            el("span", { class: "badge badge-degraded", text: "drifted" }),
            el("div", { class: "muted mono", text: `approved ${t.approved_hash}` }),
            el("div", { class: "muted mono", text: `current  ${t.current_hash}` }),
          ])
        : el("span", { class: "muted", text: t.approved_hash ? "matches pin" : "—" }),
      controls(
        button(t.disabled ? "Enable" : "Disable", t.disabled ? "btn" : "btn btn-deny", () =>
          void toggle(t),
        ),
        button("Override", "btn", () => form.show(overrideForm(t))),
      ),
    ];
  }

  async function draw(): Promise<void> {
    if (!root) return;
    let toolErr: unknown = null;
    let quarErr: unknown = null;
    try {
      const answer = await hub.listTools(filter);
      tools = { ...answer, tools: answer.tools ?? [] };
    } catch (err) {
      tools = null;
      toolErr = err;
    }
    try {
      const answer = await hub.listQuarantine();
      quarantine = { ...answer, entries: answer.entries ?? [] };
    } catch (err) {
      quarantine = null;
      quarErr = err;
    }
    clear(root);

    const server = textInput(filter, "filter by server id (empty = all)");
    const applyFilter = button("Filter", "btn", () => {
      filter = server.value.trim();
      void draw();
    });

    root.append(
      section(
        "Tool governance",
        controls(server, applyFilter),
        slot.node,
        form.node,
        toolErr
          ? failureState(toolErr, "the tool records", () => void draw())
          : (tools?.tools ?? []).length === 0
            ? empty(
                "No tool records yet.",
                "A record appears the first time a tool is seen or governed. Connect a server on the Servers page and the catalog fills itself in.",
              )
            : table(
                ["Tool", "State", "Integrity", "Actions"],
                (tools?.tools ?? []).map(toolRow),
              ),
        el("p", {
          class: "hint",
          text: "Records are keyed by the ORIGINAL downstream tool name, so a rename never moves a rule out from under itself.",
        }),
      ),
      section(
        "Quarantine",
        quarErr
          ? failureState(quarErr, "the quarantine list", () => void draw())
          : quarantineAlert(quarantine?.entries ?? []),
        quarErr
          ? null
          : (quarantine?.entries ?? []).length === 0
            ? null
            : table(
                ["Exposed name", "Origin", "Reason", "Since", ""],
                (quarantine?.entries ?? []).map((q) => [
                  el("strong", { text: q.exposed }),
                  el("div", {}, [
                    el("div", { text: `${q.server}/${q.tool}` }),
                    q.pinned_hash
                      ? el("div", { class: "muted mono", text: `pinned ${q.pinned_hash}` })
                      : null,
                    q.current_hash
                      ? el("div", { class: "muted mono", text: `current ${q.current_hash}` })
                      : null,
                  ]),
                  el("span", { text: q.reason || "—" }),
                  el("span", { text: relTime(q.at) }),
                  button("Re-approve", "btn btn-deny", () => void release(q.exposed)),
                ]),
              ),
        el("p", {
          class: "hint",
          text: "The quarantine is keyed by the CLIENT-VISIBLE name: that is what an agent could have called.",
        }),
      ),
    );
  }

  return {
    render(node) {
      root = node;
      return draw();
    },
    dispose() {
      root = null;
      form.hide();
    },
  };
}
